unixfs/tar/writer.go

```diff
@@ -61,91 +64,127 @@ func DagArchive(ctx cxt.Context, nd *mdag.Node, name string, dag mdag.DAGService
+			_, err := dagr.WriteTo(w.w)
 			return err
 		}
 
-		npath := path.Join(fpath, nd.Links[i].Name)
-		if err := w.WriteNode(ctx, child, npath); err != nil {
+		w.HandleDir = func(fpath string) error {
+			return uio.ErrIsDir
+		}
+	} else {
+		// tar and maybe gz case
+		// construct the tar writer
+		w.TarW = tar.NewWriter(w.w)
+
+		w.HandleFile = func(nd *mdag.Node, pb *upb.Data, fpath string) error {
+			if err := writeFileHeader(w.TarW, fpath, pb.GetFilesize()); err != nil {
+				return err
+			}
+
+			dagr := uio.NewDataFileReader(w.ctx, nd, pb, w.Dag)
+			_, err := dagr.WriteTo(w.TarW)
 			return err
 		}
-	}
 
-	return nil
-}
+		w.HandleDir = func(fpath string) error {
+			return writeDirHeader(w.TarW, fpath)
+		}
 
-func (w *Writer) WriteFile(ctx cxt.Context, nd *mdag.Node, fpath string) error {
-	pb := new(upb.Data)
-	if err := proto.Unmarshal(nd.Data, pb); err != nil {
-		return err
+		w.Close = w.TarW.Close
 	}
 
-	return w.writeFile(ctx, nd, pb, fpath)
+	return w, nil
 }
 
-func (w *Writer) writeFile(ctx cxt.Context, nd *mdag.Node, pb *upb.Data, fpath string) error {
-	if err := writeFileHeader(w.TarW, fpath, pb.GetFilesize()); err != nil {
-		return err
+func (w *Writer) writeDir(nd *mdag.Node, fpath string) error {
+	if w.HandleDir != nil {
+		if err := w.HandleDir(fpath); err != nil {
+			return err
+		}
 	}
 
-	dagr, err := uio.NewDagReader(ctx, nd, w.Dag)
-	if err != nil {
-		return err
-	}
+	for i, ng := range w.Dag.GetDAG(w.ctx, nd) {
+		child, err := ng.Get(w.ctx)
+		if err != nil {
+			return err
+		}
 
-	_, err = io.Copy(w.TarW, dagr)
-	if err != nil && err != io.EOF {
-		return err
+		npath := path.Join(fpath, nd.Links[i].Name)
+		if err := w.NodeWalk(child, npath); err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-func (w *Writer) WriteNode(ctx cxt.Context, nd *mdag.Node, fpath string) error {
+func (w *Writer) NodeWalk(nd *mdag.Node, fpath string) error {
 	pb := new(upb.Data)
 	if err := proto.Unmarshal(nd.Data, pb); err != nil {
 		return err
 	}
 
 	switch pb.GetType() {
+	case upb.Data_Metadata:
+		fallthrough
 	case upb.Data_Directory:
-		return w.WriteDir(ctx, nd, fpath)
+		return w.writeDir(nd, fpath)
+	case upb.Data_Raw:
+		fallthrough
 	case upb.Data_File:
-		return w.writeFile(ctx, nd, pb, fpath)
+		return w.HandleFile(nd, pb, fpath)
 	default:
-		return fmt.Errorf("unixfs type not supported: %s", pb.GetType())
+		return ft.ErrUnrecognizedType
 	}
 }
 
-func (w *Writer) Close() error {
-	return w.TarW.Close()
+func GetTarSize(ctx cxt.Context, nd *mdag.Node, dag mdag.DAGService) (uint64, error) {
+	size := 2 * uint64(tarBlockSize) // tar root padding
+
+	w := &Writer{
+		Dag: dag,
+		ctx: ctx,
+	}
+
+	w.HandleFile = func(nd *mdag.Node, pb *upb.Data, fpath string) error {
+		unixSize := pb.GetFilesize()
+		size += tarBlockSize + unixSize + (tarBlockSize - unixSize%tarBlockSize)
+		return nil
+	}
+
+	if err := w.NodeWalk(nd, ""); err != nil {
+		return 0, err
+	}
+	return size, nil
```

Comment:
is this just for the progress bar? i dont think forcing the sender to look through the entire dag before sending a size is a good idea. it might be _huge_ and take many minutes to compute just the size.

if there's no other way, perhaps the progress bar can estimate and make it clear it's an estimate of the size. as long as it doesn't look borked (i.e. negative numbers, or `bigger/smaller`) and we say "approx X MB" or "est X MB" or even "~X MB" it can be ok.

(relatedly, @krl had suggested calculating a lot of information about dags locally and keeping it around, things like depth, # of nodes, fanout, etc. but i think that's too far off for this. approx result may be better).